### PR TITLE
Detect permissions issues running flutter on Windows

### PR DESCRIPTION
**Pull request by DanTup**
I think this is a reasonable fix for #17972 (/ #17967 / #17742). Currently Windows users just get stuck in an infinite loop if they put Flutter in a folder they don't have write access to (such as `Program Files` and then run as non-Admin).

### CODE REVIEW

**Comment by devoncarew** [comment]
@goderbauer, are you the best one to review this PR?

**Comment by devoncarew** [comment]
@DanTup, @mit-mit, what about updating the getting started page for windows to specifically call out not installing flutter into `Program Files`?

**Comment by mit-mit** [comment]
Sure, but there are many other folders with similar issues.

**Comment by DanTup** [comment]
I think there could be value calling it out given how common it seems to be (probably because people are used to install apps there). I think `Program Files`, `Program Files (x86)` and `Windows` are the main ones (I can't find a full list on the MS site though). It's better to avoid the issue altogether than have someone put it in the wrong place and have to move it?

File: bin/flutter.bat [hunk 1]
```diff
@@ -47,6 +47,18 @@ IF NOT EXIST "%flutter_root%\.git" (
...
 IF NOT EXIST "%cache_dir%" MKDIR "%cache_dir%"
 
+REM If the cache still doesn't exist, fail with an error that we probably don't have permissions.
+IF NOT EXIST "%cache_dir%" (
+  ECHO Error: Unable to create cache directory at
+  ECHO            %cache_dir%
+  ECHO.
+  ECHO        This may because you do not have sufficient permissions to write to
+  ECHO        this folder and is often caused by installing Flutter into a
+  ECHO        folder like C:\Program Files and then running as a non-Admin.
```
Comment (goderbauer) [inline bin/flutter.bat]: This may encourage users to run flutter ad admin, which is something we shouldn't do IMHO. Actually, on Linux we explicitly disallow that. Can we rephrase this to something that doesn't push users towards running things as admin? E.g. "This may be because flutter doesn't have write permissions for %FLUTTER_ROOT%. Consider moving flutter to a writable path, e.g. your home directory."

**Comment by goderbauer** [comment]
+1 to warning people about extracting Flutter into directories that are known to be problematic.

It might be even better to give users an explicit example path in the instructions that they should use. The instructions currently say:

> Extract the zip-file and place the contained flutter in the desired installation location for the Flutter SDK
 
This could be changed to: "Extract the zip-file and place the contained flutter directory in the desired installation location for the Flutter SDK (e.g. `C:\src\flutter`)"

File: bin/flutter.bat [hunk 2]
```diff
@@ -47,6 +47,18 @@ IF NOT EXIST "%flutter_root%\.git" (
 REM Ensure that bin/cache exists.
 IF NOT EXIST "%cache_dir%" MKDIR "%cache_dir%"
 
+REM If the cache still doesn't exist, fail with an error that we probably don't have permissions.
+IF NOT EXIST "%cache_dir%" (
+  ECHO Error: Unable to create cache directory at
+  ECHO            %cache_dir%
+  ECHO.
+  ECHO        This may be because flutter doesn't have write permissions for
+  ECHO        this path. Try moving flutter to a writable folder, such as
```
Comment (goderbauer) [inline bin/flutter.bat]: nit: maybe say "try moving the flutter **directory**..." since we don't want people to just move the flutter executable?

**Review by goderbauer — APPROVED** [review]
LGTM

File: bin/flutter.bat (hunk 2 above)
Comment (DanTup) [reply bin/flutter.bat]: Done. I've also changed the next instance of `folder` to `location` since it seemed weird to mix directory and folder and if I made them all the same there were three instances of `directory` in that line (move your flutter directory to a writeable directory such as your home directory).

Now reads:

> This may be because flutter doesn't have write permissions for
> this path. Try moving the flutter directory to a writable location,
> such as within your home directory.

I'll land this, but feel free to request changes (or just change it) if you're not sold on it :)

**Comment by DanTup** [comment]
> Extract the zip-file and place the contained flutter directory in the desired installation location for the Flutter SDK (e.g. `C:\src\flutter`)

I was going to add to this:

> Do not extract Flutter to a folder that requires elevated permissions (like `C:\Program Files\`).

However this could be misinterpreted as that being a suggestion of somewhere good. Is the first change above enough, or should we include something like the second line (but maybe word it in a less ambiguous way)?
